traits/trait_types.py

```diff
@@ -3336,6 +3338,144 @@ def as_ctrait(self):
 
+# -------------------------------------------------------------------------------
+#  'None' trait:
+# -------------------------------------------------------------------------------
+
+
+class NoneTrait(TraitType):
+    """ Defines a trait that only accepts the None value
+    """
+
+    info_text = "None"
+
+    default_value = None
+
+    default_value_type = DefaultValue.constant
+
+    def __init__(self, **metadata):
+        default_value = metadata.pop("default", None)
+        if default_value is not None:
+            raise ValueError("Cannot set default value {} "
+                             "for NoneTrait".format(default_value))
+        super(NoneTrait, self).__init__(**metadata)
+
+    def validate(self, obj, name, value):
+        if value is None:
+            return value
+
+        self.error(obj, name, value)
+
+# -------------------------------------------------------------------------------
+#  'Union' trait:
+# -------------------------------------------------------------------------------
+
+
+class Union(TraitType):
+    """ Defines a trait whose value can be any of of a specified list of
+    trait types or list of trait type instances or None
+    """
+    # CTrait type map for special trait types:
+    type_map = {"event": TraitKind.event, "constant": TraitKind.constant}
+
+    def __init__(self, *traits, **metadata):
+        self.list_ctrait_instances = []
+        traits = traits or [NoneTrait]
+        for trait in traits:
+            if trait is None:
+                trait = NoneTrait
+            ctrait_instance = trait_cast(trait)
+            if ctrait_instance is None:
+                raise ValueError("Union trait declaration expects a trait "
+                                 "type or an instance of trait type or None,"
+                                 " but got {} instead".format(trait))
+
+            self.list_ctrait_instances.append(ctrait_instance)
+
+        default_value = None
+        if 'default' in metadata:
+            default_value = metadata.pop("default")
+        elif self.list_ctrait_instances:
+            default_value = self.list_ctrait_instances[0].default
+
+        self.default_value_type = _infer_default_value_type(default_value)
+        super().__init__(default_value, **metadata)
+
+    def validate(self, obj, name, value):
+        """ Return the value by the first trait in the list that can
+        validate the assigned value, raise an error if none of them can.
+        """
+        for trait_type_instance in self.list_ctrait_instances:
+            try:
+                return trait_type_instance.validate(obj, name, value)
+            except TraitError:
+                pass
+
+        self.error(obj, name, value)
+
+    def info(self):
+        return " or ".join([ctrait.info() for ctrait in
+                            self.list_ctrait_instances])
+
+    def inner_traits(self):
+        return tuple(self.list_ctrait_instances)
+
+    def as_ctrait(self):
```

Comment:
A lot of this code is common with `TraitType.as_ctrait`.  At first inspection it _looks_ like you won't need to override this at all, but I could be wrong.

If I am wrong and there is something that this does differently, is there a small change that coule be made to `TraitType.as_ctrait` to cover the case?

Reply:
Turns out it is indeed, which is great! Even sets the handler correctly, so I just removed the overridden method.